src/Features/LanguageServer/Protocol/Features/CodeCleanup/AbstractCodeCleanupService.cs

```diff
@@ -162,7 +189,80 @@ private async Task<Document> ApplyCodeFixesForSpecificDiagnosticIdAsync(Document
             return solution.GetDocument(document.Id) ?? throw new NotSupportedException(FeaturesResources.Removal_of_document_not_supported);
         }
 
+        private async Task<ImmutableArray<(string diagnosticId, string? title)>> GetThirdPartyDiagnosticIdsAndTitlesAsync(Document document, CancellationToken cancellationToken)
+        {
+            var tree = await document.GetRequiredSyntaxTreeAsync(cancellationToken).ConfigureAwait(false);
+            var range = new TextSpan(0, tree.Length);
+
+            // Compute diagnostics for everything that is not an IDE analyzer
+            var diagnostics = (await _diagnosticService.GetDiagnosticsForSpanAsync(document, range,
+                shouldIncludeDiagnostic: static diagnosticId => !(IDEDiagnosticIdToOptionMappingHelper.IsKnownIDEDiagnosticId(diagnosticId)),
+                includeCompilerDiagnostics: true, includeSuppressedDiagnostics: false,
+                cancellationToken: cancellationToken).ConfigureAwait(false));
+
+            // ensure more than just known diagnostics were returned
+            if (!diagnostics.Any())
+            {
+                return ImmutableArray<(string diagnosticId, string? title)>.Empty;
+            }
+
+            return diagnostics.SelectAsArray(static d => (d.Id, d.Title)).Distinct();
+        }
+
+        private async Task<Document> ApplyThirdPartyCodeFixesAsync(
+            Document document,
+            ImmutableArray<(string diagnosticId, string? title)> diagnosticIds,
+            IProgressTracker progressTracker,
+            CodeActionOptionsProvider fallbackOptions,
+            CancellationToken cancellationToken)
+        {
+            foreach (var (diagnosticId, title) in diagnosticIds)
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+
+                progressTracker.Description = string.Format(FeaturesResources.Fixing_0, title ?? diagnosticId);
+                // Apply codefixes for diagnostics with a severity of warning or higher
+                var updatedDocument = await _codeFixService.ApplyCodeFixesForSpecificDiagnosticIdAsync(
+                    document, diagnosticId, DiagnosticSeverity.Warning, progressTracker, fallbackOptions, cancellationToken).ConfigureAwait(false);
+
+                // If changes were made to the solution snap shot outside the current document discard the changes.
+                // The assumption here is that if we are applying a third party code fix to a document it only affects the document.
+                // Symbol renames and other complex refactorings we do not want to include in code cleanup.
```

Comment:
Can we do a follow up item to potentially log telemetry when this happens and notify the user?